test/solve_dense.cpp

```diff
@@ -12,9 +12,163 @@
 // issue https://github.com/arrayfire/arrayfire/issues/1617
 
 #include <gtest/gtest.h>
+
 #include <testHelpers.hpp>
+#include <af/algorithm.h>
+#include <af/arith.h>
+#include <af/blas.h>
+#include <af/defines.h>
+#include <af/device.h>
+#include <af/dim4.hpp>
+#include <af/lapack.h>
+#include <af/traits.hpp>
+
+#include <complex>
+#include <iostream>
+#include <string>
 #include <thread>
-#include "solve_common.hpp"
+#include <vector>
+
+using af::array;
+using af::cdouble;
+using af::cfloat;
+using af::deviceGC;
+using af::dim4;
+using af::dtype_traits;
+using af::setDevice;
+using af::sum;
+using std::abs;
+using std::cout;
+using std::endl;
+using std::string;
+using std::vector;
+
+template<typename T>
+void solveTester(const int m, const int n, const int k, const int b, double eps,
+                 int targetDevice = -1) {
+    if (targetDevice >= 0) setDevice(targetDevice);
+
+    deviceGC();
+
+    SUPPORTED_TYPE_CHECK(T);
+    if (noLAPACKTests()) return;
+
+#if 1
+    array A  = cpu_randu<T>(dim4(m, n, b));
+    array X0 = cpu_randu<T>(dim4(n, k, b));
+#else
+    array A  = randu(m, n, (dtype)dtype_traits<T>::af_type);
+    array X0 = randu(n, k, (dtype)dtype_traits<T>::af_type);
+#endif
+    array B0 = matmul(A, X0);
+
+    //! [ex_solve]
+    array X1 = solve(A, B0);
+    //! [ex_solve]
+
+    //! [ex_solve_recon]
+    array B1 = matmul(A, X1);
+    //! [ex_solve_recon]
+
+    ASSERT_NEAR(
+        0,
+        sum<typename dtype_traits<T>::base_type>(abs(real(B0 - B1))) / (m * k),
+        eps);
+    ASSERT_NEAR(
+        0,
+        sum<typename dtype_traits<T>::base_type>(abs(imag(B0 - B1))) / (m * k),
+        eps);
```

Comment:
I have a local change where I am checking why we don't use `ASSERT_ARRAYS_NEAR` in solve tests. Are you facing any issues using those macros with solve ? because I do, hence I didn't go ahead with my change - https://github.com/9prady9/arrayfire/commit/a683c8858e64ab9d22a0a2af7dfe9185cd64cd0b

Reply:
I picked your commit.